File: sw_x/gvm_core/internal/protorpc/csharp/ProtoRpcClient.cs
//
//  Copyright 2011 iGware Inc.
//  All Rights Reserved.
//
//  THIS SOFTWARE CONTAINS CONFIDENTIAL INFORMATION AND
//  TRADE SECRETS OF IGWARE INC.
//  USE, DISCLOSURE OR REPRODUCTION IS PROHIBITED WITHOUT
//  THE PRIOR EXPRESS WRITTEN PERMISSION OF IGWARE INC.
//
using System;
using System.IO;
using rpc;

namespace protorpc
{
    public class ProtoRpcClient
    {
        private ProtoChannel _channel;
        private bool _throwOnAppError;

        /**
         * Allows communicating with a ProtoRPC-based service over a {@link ProtoChannel}.
         * @param channel Provide an implementation of {@link ProtoChannel} that connects
         *          to the target service.
         * @param throwOnAppError If <code>true</code>, then
         *          {@link #sendRpc(String, MessageLite, com.google.protobuf.MessageLite.Builder)}
         *          and {@link #recvRpcResponse(com.google.protobuf.MessageLite.Builder)}
         *          will throw {@link AppLayerException} instead of returning a negative
         *          app-layer error code (since the response will be empty in this case anyway).
         */
        public ProtoRpcClient(ProtoChannel channel, bool throwOnAppError)
        {
            _channel = channel;
            _throwOnAppError = throwOnAppError;
        }
        
        /**
         * Sends an RPC over the channel.  The RPC is dispatched to the named
         * method.  Upon return, a valid status will be filled in, and the
         * response may or may not be filled in.  The caller should check the
         * status to see if the call was successful.  The response will only be
         * filled in when (RpcStatus.getStatus() == Status.OK and RpcStatus.getAppStatus() >= 0).
         * @throws ProtoRpcException 
         */
        protected int sendRpc(string methodName, object request, Type type, ref object response)
        {
            sendRpcRequest(methodName, request);
            return recvRpcResponse(type, ref response);
        }
        
        // sendRpc() split into two halves, for async support:
        
        protected void sendRpcRequest(string methodName, object request)
        {
            try {
                // Send request header
                RpcRequestHeader reqHdr = new RpcRequestHeader();
                reqHdr.methodName = methodName;
                _channel.writeMessage(reqHdr);
                
                // Send request body
                _channel.writeMessage(request);
                
                // Flush stream
                _channel.flushOutputStream();
                
            } catch (RpcLayerException e) {
                throw e;
            } catch (IOException e) {
                throw new RpcLayerException(RpcStatus.Status.IO_ERROR, e.Message);
            } catch (ProtoBuf.ProtoException e) {
                throw new RpcLayerException(RpcStatus.Status.BAD_REQUEST, e.Message);
            } catch (Exception e) {
                throw new RpcLayerException(RpcStatus.Status.INTERNAL_ERROR, e.Message);
            }
        }

        protected int recvRpcResponse(Type type, ref object response)
        {
            try {
                RpcStatus status = new RpcStatus();
                object reference = status;
                _channel.extractMessage(status.GetType(), ref reference);
                status = (RpcStatus)reference;
                if (status.status != RpcStatus.Status.OK)
                {
                    throw new RpcLayerException(status.status, status.errorDetail);
                }
                if (status.appStatus >= 0) {
                    _channel.extractMessage(type, ref response);
                } else if (_throwOnAppError) {
                    throw new AppLayerException(status.appStatus);
                }
                return status.appStatus;
                
            } catch (ProtoRpcException e) {
                throw e;
            } catch (IOException e) {
                throw new RpcLayerException(RpcStatus.Status.IO_ERROR, e.Message);
            } catch (ProtoBuf.ProtoException e) {
                throw new RpcLayerException(RpcStatus.Status.BAD_RESPONSE, e.Message);
            } catch (Exception e) {
                throw new RpcLayerException(RpcStatus.Status.INTERNAL_ERROR, e.Message);
            }
        }
    }
}

File: sw_x/gvm_core/internal/protorpc/ProtoRpcClient.cpp
/*
 *               Copyright (C) 2010, BroadOn Communications Corp.
 *
 *  These coded instructions, statements, and computer programs contain
 *  unpublished  proprietary information of BroadOn Communications Corp.,
 *  and  are protected by Federal copyright law. They may not be disclosed
 *  to  third  parties or copied or duplicated in any form, in whole or in
 *  part, without the prior written consent of BroadOn Communications Corp.
 *
 */

#include "ProtoRpcClient.h"

ProtoRpcClient::ProtoRpcClient(ProtoChannel& channel)
    : _channel(channel)
{}


bool
ProtoRpcClient::sendRpcRequest(
        const std::string& methodName,
        const ProtoMessage& request,
        RpcStatus& status)
{
    // Check for previous IO errors
    if (_channel.outputStreamError()) {
        status.set_status(RpcStatus_Status_IO_ERROR);
        status.set_errordetail("IO error occurred previously on output stream");
        goto fail;
    }
    if (_channel.inputStreamError()) {
        status.set_status(RpcStatus_Status_IO_ERROR);
        status.set_errordetail("IO error occurred previously on input stream");
        goto fail;
    }

    // Send request header
    {
        RpcRequestHeader reqHdr;
        reqHdr.set_methodname(methodName);
        if (!_channel.writeMessage(reqHdr, NULL)) {
            if (_channel.outputStreamError()) {
                goto output_stream_error;
            } else {
                status.set_status(RpcStatus_Status_INTERNAL_ERROR);
                status.set_errordetail("Problem sending request header: ");
                status.mutable_errordetail()->append(reqHdr.InitializationErrorString());
                goto fail;
            }
        }
    }

    // Send request body
    if (!_channel.writeMessage(request, NULL)) {
        if (_channel.outputStreamError()) {
            goto output_stream_error;
        } else {
            status.set_status(RpcStatus_Status_BAD_REQUEST);
            status.set_errordetail("Missing request field(s): ");
            status.mutable_errordetail()->append(request.InitializationErrorString());
            goto fail;
        }
    }

    // Flush stream
    if (!_channel.flushOutputStream()) {
        goto output_stream_error;
    }

    // Everything OK so far.
    return true;

output_stream_error:
    {
        status.set_status(RpcStatus_Status_IO_ERROR);
        std::string errorDetail;
        _channel.outputStreamErrorDetail(errorDetail);
        status.set_errordetail(errorDetail);
    }

fail:
    return false;
}

void
ProtoRpcClient::recvRpcResponse(ProtoMessage& response, RpcStatus& status)
{
    // Get response header
    if (!_channel.extractMessage(status)) {
        if (_channel.inputStreamError()) {
            status.Clear();
            goto input_stream_error;
        } else {
            std::string error("Problem receiving response header: ");
            error.append(status.InitializationErrorString());
            status.Clear();
            status.set_status(RpcStatus_Status_HEADER_ERROR);
            status.set_errordetail(error);
            goto done;
        }
    }

    // Get response body if status OK
    if ((status.status() == RpcStatus_Status_OK) && (status.appstatus() >= 0)) {
        if (!_channel.extractMessage(response)) {
            if (_channel.inputStreamError()) {
                status.Clear();
                goto input_stream_error;
            } else {
                status.set_status(RpcStatus_Status_BAD_RESPONSE);
                status.set_errordetail(response.InitializationErrorString());
                goto done;
            }
        }
    }

    // Everything OK.  Status was parsed from channel.
    goto done;

input_stream_error:
    {
        status.set_status(RpcStatus_Status_IO_ERROR);
        std::string errorDetail;
        _channel.inputStreamErrorDetail(errorDetail);
        status.set_errordetail(errorDetail);
    }
    goto done;

done:
    NO_OP();
}

void
ProtoRpcClient::sendRpc(const std::string& methodName,
        const ProtoMessage& request, ProtoMessage& response,
        RpcStatus& status)
{
    bool success = sendRpcRequest(methodName, request, status);
    if (success) {
        response.Clear();
        recvRpcResponse(response, status);
    }
}

const ProtoChannel&
ProtoRpcClient::channel(void)
{
    return _channel;
}
